express-notes-api/index.js

```diff
@@ -0,0 +1,102 @@
+const express = require('express');
+const fs = require('fs');
+const data = require('./data.json');
+const app = express();
+const jsonMiddleware = express.json();
+app.use(jsonMiddleware);
+
+app.get('/api/notes', (req, res, next) => {
+  const dataNotes = [];
+  for (const object in data.notes) {
+    dataNotes.push(data.notes[object]);
+  }
+  res.send(dataNotes);
+});
+
+app.get('/api/notes/:id', (req, res, next) => {
+  const dataNotes = [];
+  for (const object in data.notes) {
+    dataNotes.push(data.notes[object]);
+  }
+  if (req.params.id <= 0) {
+    res.status(400).send({ error: 'id must be a positive integer' });
+  }
+  if (req.params.id > dataNotes.length) {
+    res.status(404).send({ error: 'cannot find note with id ' + req.params.id });
+  }
+  const singleNote = data.notes[req.params.id];
+  res.send(singleNote);
```

Comment:
Ok, this is problematic because it is technically possible for the server to attempt 2 responses to the same request. `res.send()` does not exit this middleware function. Only `return` and `throw` exit a function.